src/content/blog/2024-05-09-eslint-compatibility-utilities.md

```diff
@@ -0,0 +1,88 @@
+---
+layout: post
+title: Introducing ESLint Compatibility Utilities
+teaser: "Do you rely on a plugin that doesn't yet work with ESLint v9? This package will help."
+tags:
+  - Compatibility
+  - Tools
+authors:
+  - nzakas
+categories:
+  - Announcements
+---
+
+The release of ESLint v9.0.0 brought with it the rollout of the [new configuration system](/blog/2023/10/flat-config-rollout-plans/), but also a series of changes to the [rules API](/blog/2023/09/preparing-custom-rules-eslint-v9/). These changes are necessary in order to prepare ESLint for implementing [language plugins](https://github.com/eslint/rfcs/blob/main/designs/2022-languages/README.md), which will give ESLint the ability to natively lint languages other than JavaScript. As a result, plugin authors needed to update their rules to work with v9.0.0, and unfortunately, that means some of the plugins you rely on may not have been updated yet. That's why we're releasing the compatibility utilities.
+
+## How to know if the compatibility utilities will help
+
+These utilities may help if you encounter any of the following errors while running ESLint:
+
+```
+TypeError: context.getScope is not a function
+TypeError: context.getAncestors is not a function
+TypeError: context.markVariableAsUsed is not a function
+TypeError: context.getDeclaredVariables is not a function
+```
+
+These errors mean that the plugin rules have not been updated to latest ESLint rule API.
+
+## Using the compatibility utilities
+
+First, install the `@eslint/compat` package using npm or any npm-compatible CLI:
+
+```shell
+npm install @eslint/compat -D
+# or
+yarn add @eslint/compat -D
+# or
+pnpm install @eslint/compat -D
+# or
+bun install @eslint/compat -D
+```
+
+Then, use the `fixupPluginRules()` function in your `eslint.config.js` file to wrap the plugin in a compatibility layer:
+
+```js
+// eslint.config.js
+import { fixupPluginRules } from "@eslint/compat";
+import oldPlugin from "eslint-plugin-old-plugin";
+
+export default [
+    {
+        plugins: {
+            oldPlugin: fixupPluginRules(oldPlugin)
+        }
+    },
+
+    // other config
+];
+```
+
+After that, the plugin should work as expected.
+
```

Comment:
Would it make sense to also provide an example with a plugin that exports a flat config?

For example, eslint-plugin-react exports flat configs:

https://github.com/jsx-eslint/eslint-plugin-react?tab=readme-ov-file#configuration-new-eslintconfigjs

Rules aren't compatible with the new API yet, but the config can be used like this:

```js
// eslint.config.js
import { fixupConfigRules } from "@eslint/compat";
import reactRecommended  from "eslint-plugin-react/configs/recommended.js";

export default [
    ...fixupConfigRules(reactRecommended)
];
```

Reply:
Good idea.